keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1297 @@
+in the PodResourceConfig message, similar to the container-level resource
+information.
+
+```diff
+ message PodSandboxConfig {
+ 
+ ...
+ 
+     // Optional configurations specific to Linux hosts.
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++
++    // Kubernetes resource spec of the containers in the pod.
++    PodResourceConfig pod_resources = 10;
+ }
+ 
++// PodResourceConfig contains information of all resources requirements of
++// the containers of a pod.
++message PodResourceConfig {
++    // Resource configuration of all containers in the pod.
++    repeated ContainerResourceConfig containers = 1;
++ 
++    // Kubernetes resource spec of the pod-level resource requirements.
++    // This is the pod-level resource requirements introduced in KEP-2837
++    // (alpha in v1.32).
++    KubernetesResources kubernetes_resources = 2;
++}
+ 
++// ContainerResourceConfig contains information of all resource requirements of
++// one container.
++message ContainerResourceConfig {
++    // Name of the container
++    string name= 1;
++
++    // Type of the container
++    ContainerType type= 2;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 3;
++
++    // Mounts for the container.
++    repeated Mount mounts = 4;
++
++    // Devices for the container.
++    repeated Device devices = 5;
++
++    // CDI devices for the container.
++    repeated CDIDevice CDI_devices = 6;
++}
+
++enum ContainerType {
++    INIT_CONTAINER    = 0;
++    SIDECAR_CONTAINER = 1;
++    CONTAINER = 2;
++}
+```
+
+#### CreateContainer
+
+The ContainerConfig message (used in CreateContainer request) is extended to
+contain unmodified resource requests from the PodSpec.
+
+```diff
++import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+ message ContainerConfig {
+ 
+ ...
+ 
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
+ 
+     // CDI devices for the container.
+     repeated CDIDevice CDI_devices = 17;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 18;
+ }
+ 
++// KubernetesResources contains the resource requests and limits as specified
++// in the Kubernetes core API ResourceRequirements.
++message KubernetesResources {
++    // Requests and limits from the Kubernetes container config.
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> requests = 1;
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> limits = 2;
++}
+```
+
+Note that mounts, devices, CDI devices are part of the ContainerConfig message
+but are left out of the diff snippet above.
+
+Including the KubernetesResources in the ContainerConfig message serves
+multiple purposes:
+
+1. Catch changes that happen between pod sandbox creation and container
```

Comment:
in place resize should be writing a message to the runtime that a pod or container had changed resources, right @tallclair ? I think that'd be better than having a runtime interpret diffs in container values...

Reply:
Does it do so when the container hasn't yet been created. IIRC the thinking behind this case were the kata/coco use cases. The runtime or nri plugin whatever CAN detect changes if they need/want to.

Reply:
Let's validate this flow. I don't think letting the runtime handle this infer this from diffs is correct.